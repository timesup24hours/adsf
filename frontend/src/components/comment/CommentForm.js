import React from 'react'
import TextFieldGroupMaterialUI from '../common/TextFieldGroupMaterialUI'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import * as commentActions from '../../store/actions/commentActions'
import commentValidator from './commentValidator'
import RaisedButton from 'material-ui/RaisedButton'
import Chip from 'material-ui/Chip'
import { showLoading, hideLoading } from 'react-redux-loading-bar'
import { errorToMessage } from '../../utils'

export const CommentForm = props => {
  const { comment } = props

  const onChangeField = e => {
    e.preventDefault()
    props.commentSetValue({ [e.target.name]: e.target.value })
    props.commentSetError({ [e.target.name]: undefined })
  }

  const isValid = () => {
    const { errors, isValid } = commentValidator(comment.values)
    if (!isValid) {
      props.commentSetError(errors)
      props.commentSubmitFail()
    }

    return isValid
  }

  const handleSubmit = e => {
    e.preventDefault()
    if (isValid()) {
      props.showLoading()
      props.commentSubmitRequest(comment.values)
      .then(res => {
        if(res.data.success) {
          props.commentResetValues()
          props.commentSubmitSuccess(res.data.comment)
          props.hideLoading()
        }
      })
      .catch(e => {
        console.log(e);
        props.hideLoading()
        props.commentSubmitFail()
        props.commentSetError(errorToMessage(e.response))
      })
    }
  }

  return (
    <div className='CommentForm-container'>
      {comment.error
        ? <div className='CommentForm-general-errors'>
            <Chip backgroundColor='#ff7276'>{comment.errors.login}</Chip>
          </div>
        : null
      }
      <form onSubmit={handleSubmit}>
        <TextFieldGroupMaterialUI
          className='CommentForm-comment'
          error={comment.errors.comment}
          label="Comment"
          onChange={onChangeField}
          value={comment.values.comment}
          name="comment"
          multiLine={true}
          rowsMax={4}
        />
      <div className='CommentForm-button'>
        <RaisedButton label="Comment"  type="submit" disabled={comment.pending || comment.error} />
      </div>
      </form>
    </div>
  )
}

const mapStateToProps = state => ({
  // comment: state.comment,
})

const mapDispatchToProps = dispatch => ({
  commentSubmitting: () => dispatch(commentActions.commentSubmitting()),
  commentSubmitSuccess: res => dispatch(commentActions.commentSubmitSuccess(res)),
  commentSubmitFail: () => dispatch(commentActions.commentSubmitFail()),
  commentSetValue: bindActionCreators(commentActions.commentSetValue, dispatch),
  commentSetError: error => dispatch(commentActions.commentSetError(error)),
  commentSubmitRequest: bindActionCreators(commentActions.commentSubmitRequest, dispatch),
  commentResetValues: () => dispatch(commentActions.commentResetValues()),
  showLoading: () => dispatch(showLoading()),
  hideLoading: () => dispatch(hideLoading()),
})

export default connect(mapStateToProps, mapDispatchToProps)(CommentForm)
