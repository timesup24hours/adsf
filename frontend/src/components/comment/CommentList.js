import React, { Component, PropTypes } from 'react'
import moment from 'moment'
import Delete from 'material-ui/svg-icons/action/delete-forever'
import Edit from 'material-ui/svg-icons/editor/border-color'
import Save from 'material-ui/svg-icons/content/save'
import Cancel from 'material-ui/svg-icons/navigation/cancel'
import { connect } from 'react-redux'
import * as commentActions from '../../store/actions/commentActions'
import Dialog from 'material-ui/Dialog'
import FlatButton from 'material-ui/FlatButton'
import TextFieldGroupMaterialUI from '../common/TextFieldGroupMaterialUI'
import Snackbar from 'material-ui/Snackbar'
import { snackbarClose } from '../../store/actions/snackbarActions'
import UserAvatar from '../nav/UserAvatar'

export class CommentList extends Component {
  state = {
    open: false,
    editorOpen: false,
    comment: ''
  }

  componentDidMount() {
    this.setState({ comment: this.props.comment })
  }

  handleRequestClose = () => {
    this.props.snackbarClose()
  }

  openEditor = () => {
    this.setState({ editorOpen: true })
  }

  handleOpen = () => {
    this.setState({ open: true })
  }

  handleClose = () => {
    this.setState({ open: false })
  }

  handleCancel = () => {
    this.setState({ editorOpen: false, comment: this.props.comment })
  }

  handleOnchange = e => {
    this.setState({ comment: e.target.value });
  }

  handleSave = () => {
    if(this.state.comment !== this.props.comment) {
      this.props.commentEditRequest({ id: this.props.commentId, comment: this.state.comment })
    } 
    this.setState({ editorOpen: false  })
  }

  handleDeleteComment = () => {
    this.props.commentDeleteRequest(this.props.commentId)
    this.handleClose()
  }

  render() {
    const { user, loginUserId, createdAt } = this.props
    const actions = [
      <FlatButton
        label="Cancel"
        primary={true}
        onTouchTap={this.handleClose}
      />,
      <FlatButton
        label="Delete"
        primary={true}
        keyboardFocused={true}
        onTouchTap={this.handleDeleteComment}
      />,
    ]

    return (
      <div className='CommentList-container'>
        <div className='CommentList-user-holder'>
          <div className='CommentList-user-group'>
            <div className='CommentList-avatar'>
              <UserAvatar user={user} />
            </div>
            <div className='CommentList-username-time-group'>
              <div className='CommentList-username'>{user.local.nickname || user.local.username}</div>
              <div className='CommentList-time'>{moment(createdAt).fromNow()}</div>
            </div>
          </div>
        </div>
        <div className='CommentList-comment'>
          { this.state.editorOpen
            ? <div className='CommentList-edit-input'>
                <TextFieldGroupMaterialUI
                  id='CommentList-edit-input'
                  name='comment'
                  value={this.state.comment}
                  type='text'
                  onChange={this.handleOnchange}
                  multiLine={true}
                  rows={4}
                  rowsMax={4}
                  fullWidth={true}
                />
              </div>
            : <div className='CommentList-comment-show'>{this.state.comment}</div>
          }
        </div>


          <Dialog
            title="Delete Confirm"
            actions={actions}
            modal={false}
            open={this.state.open}
            onRequestClose={this.handleClose}
          >
            This comment would be deleted permanently.
          </Dialog>


        {loginUserId && loginUserId === user._id && this.state.editorOpen
          ? <div className='CommentList-save' onTouchTap={this.handleSave} key='CommentList-save'>
              <div><Save/></div>
            </div>
          : null}
        {loginUserId && loginUserId === user._id && !this.state.editorOpen
          ? <div id='CommentList-edit' className='CommentList-edit' onTouchTap={this.openEditor} key='CommentList-edit'>
              <div><Edit/></div>
            </div>
          : null}
        {loginUserId && loginUserId === user._id && !this.state.editorOpen
          ? <div className='CommentList-delete' onTouchTap={this.handleOpen} key='CommentList-delete'>
              <div><Delete/></div>
            </div>
          : null}
        {loginUserId && loginUserId === user._id && this.state.editorOpen
          ? <div className='CommentList-cancel' onTouchTap={this.handleCancel} key='CommentList-cancel'>
              <div><Cancel/></div>
            </div>
          : null}

          <Snackbar
            open={this.props.snackbar.open}
            message={this.props.snackbar.message}
            autoHideDuration={5000}
            onRequestClose={this.handleRequestClose}
          />

      </div>
    )
  }

}

CommentList.propTypes = {
  user: PropTypes.object.isRequired,
  comment: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  commentId: PropTypes.string.isRequired
}

const mapDispatchToProps = dispatch => ({
  commentDeleteRequest: commentId => dispatch(commentActions.commentDeleteRequest(commentId)),
  commentEditRequest: paypload => dispatch(commentActions.commentEditRequest(paypload)),
  snackbarClose: () => dispatch(snackbarClose())
})

const mapStateToProps = state => ({
  comments: state.comment,
  snackbar: state.snackbar,
  auth: state.auth
})

export default connect(mapStateToProps, mapDispatchToProps)(CommentList)
