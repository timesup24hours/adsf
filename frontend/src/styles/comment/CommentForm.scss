.CommentForm-container {
  width: auto;
  display: flex;
  justify-content: center;
  height: 30vh;
  margin-top: 50px;
  flex-direction: column;
  align-items: center;
}
.CommentForm-button {
  display: flex;
  justify-content: flex-end;
}
.CommentForm-general-errors {

}
