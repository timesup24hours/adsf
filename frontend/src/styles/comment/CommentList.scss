.CommentList-container {
  margin: 0px auto 0px;
  display: flex;
  justify-content: flex-start;
  flex-direction: row;
  width: 500px;
  padding: 12px 10px;
  border-bottom: 1px solid rgba(0, 0, 0, .1);
  position: relative;

  &:hover {
    // background-color: rgba(0, 0, 0, .1);
    .CommentList-delete, .CommentList-edit {
      opacity: 0.4;
      &:hover {
        opacity: 1;
      }
    }
  }
}
.CommentList-user-holder {

}

.CommentList-user-group {
  display: flex;
  flex-direction: row;
  flex: 1;
  align-items: center;
}

.CommentList-comment {
  margin-right: 14px;
  flex: 10;
  div {
    white-space: pre-wrap;
    word-break: break-all;
  }
  .CommentList-edit-input {
    margin-top: -14px;
  }
  .CommentList-comment-show {
    // margin-top: 14px;
  }
}


.CommentList-avatar {
}

.CommentList-username-time-group {
  margin: 0 6px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  width: 80px
}
.CommentList-username {
  font-weight: lighter;
  font-size: 14px;

}
.CommentList-time {
  font-size: 10px;
  color: grey;
}

.CommentList-delete, .CommentList-edit {
  cursor: pointer;
  opacity: 0;
}
.CommentList-save, .CommentList-cancel {
  cursor: pointer;
  opacity: 0.7;
  &:hover {
    opacity: 1;
  }
}

@media screen and (max-width: 700px) {
  .CommentList-container {
    width: auto;
    padding: 10px 10px 4px;
    flex-direction: column;
  }
  .CommentList-username-time-group {
  }
  .CommentList-time {
    color: grey;
  }
  .CommentList-comment {
    margin-left: 14px;
    padding: 8px 8px;
    width: auto;
  }
  .CommentList-avatar {
  }
  .CommentList-save {
    position: absolute;
    right: 40px;
  }
  .CommentList-edit {
    position: absolute;
    right: 40px;
  }
  .CommentList-delete {
    position: absolute;
    right: 10px;
  }
  .CommentList-cancel {
    position: absolute;
    right: 10px;
  }
}
