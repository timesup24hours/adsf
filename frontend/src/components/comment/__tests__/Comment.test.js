import CommentWrapper, { Comment } from '../Comment'
import CommentList from '../CommentList'
import configureMockStore from 'redux-mock-store'
import ReactCSSTransitionGroup from 'react-addons-css-transition-group'
import CommentForm from '../CommentForm'
const mockStore = configureMockStore();

const auth = {
  user: {
    _id: 'testUserId123',
  }
}

const commentObj = {
  _id: '1',
  user: {},
  createdAt: '2099',
  comment: 'testComment',
}

const comment = {
  error: false,
  pending: false,
  success: false,
  fail: false,
  comments: [ commentObj, commentObj ],
}

const store = mockStore({
  auth,
  comment,
})

const commentGetAllRequest = jest.fn()

test('# CommentWrapper', () => {
  const wrapper = shallow(<CommentWrapper store={store} />)
  expect(wrapper).toMatchSnapshot()
})

test('# Comment', () => {
  const component = shallow(<Comment comment={{ comments: [ commentObj, commentObj ] }} auth={auth} commentGetAllRequest={commentGetAllRequest}  />)
  expect(component).toMatchSnapshot()
})

test('# Comment render nested component', () => {
  const component = shallow(<Comment comment={{ comments: [] }} auth={auth} commentGetAllRequest={commentGetAllRequest}  />)
  expect(component).toMatchSnapshot()
  expect(component.contains(<div className='Comment-noComment'>No Comment!</div>)).toEqual(true);
})

test('# Comment should run componentDidMount', () => {
  const component = (<Comment comment={comment} auth={auth} commentGetAllRequest={commentGetAllRequest} store={store}  />)
  // const app = mount(component)
})
