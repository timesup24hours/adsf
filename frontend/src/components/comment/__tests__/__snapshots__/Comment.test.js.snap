exports[`test # Comment 1`] = `
<div
  className="Comment-container">
  <Connect(Component)
    comment={
      Object {
        "comments": Array [
          Object {
            "_id": "1",
            "comment": "testComment",
            "createdAt": "2099",
            "user": Object {},
          },
          Object {
            "_id": "1",
            "comment": "testComment",
            "createdAt": "2099",
            "user": Object {},
          },
        ],
      }
    } />
  <ReactCSSTransitionGroup
    transitionAppear={false}
    transitionEnter={true}
    transitionEnterTimeout={1000}
    transitionLeave={true}
    transitionLeaveTimeout={800}
    transitionName="CommentAnimation">
    <Connect(CommentList)
      comment="testComment"
      commentId="1"
      createdAt="2099"
      loginUserId="testUserId123"
      user={Object {}} />
    <Connect(CommentList)
      comment="testComment"
      commentId="1"
      createdAt="2099"
      loginUserId="testUserId123"
      user={Object {}} />
  </ReactCSSTransitionGroup>
</div>
`;

exports[`test # Comment render nested component 1`] = `
<div
  className="Comment-container">
  <Connect(Component)
    comment={
      Object {
        "comments": Array [],
      }
    } />
  <div
    className="Comment-noComment">
    No Comment!
  </div>
</div>
`;

exports[`test # CommentWrapper 1`] = `
<Comment
  auth={
    Object {
      "user": Object {
        "_id": "testUserId123",
      },
    }
  }
  comment={
    Object {
      "comments": Array [
        Object {
          "_id": "1",
          "comment": "testComment",
          "createdAt": "2099",
          "user": Object {},
        },
        Object {
          "_id": "1",
          "comment": "testComment",
          "createdAt": "2099",
          "user": Object {},
        },
      ],
      "error": false,
      "fail": false,
      "pending": false,
      "success": false,
    }
  }
  commentGetAllRequest={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  } />
`;
